src/azul/plugins/repository/canned/__init__.py

```diff
@@ -69,7 +71,104 @@
 log = logging.getLogger(__name__)
 
 
-class CannedSourceRef(SourceRef[SimpleSourceSpec, 'CannedSourceRef']):
+@attrs.frozen(kw_only=True)
+class CannedSourceSpec(SourceSpec):
+    url: str
+    ref: str
+    path: tuple[str]
+
+    @classmethod
+    def parse(cls, spec: str) -> 'CannedSourceSpec':
```

Comment:
Now that we're on python 3.11, we can write
```suggestion
    def parse(cls, spec: str) -> Self:
```
as long as it used consistently throughout the module.